### DISCUSSION TITLE
"Preventing access to file outside the base directory" errors

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#17996] [Suggest an Idea]
Seeing a lot of log warnings bubbling up into the hosted app's monitoring like this:

```
baseDir	/mnt/renovate/gh/THEORG/layoutlib
path /mnt/renovate/gh/THEORG/common_libs/config.gradle
repository THEORG/layoutlib
```

Is this type of error something which the user could prevent through config? If not then we should not be log warning, as it will end up with messages in their PRs and Dependency Dashboard which we can't do anything about. I'm guessing this is because:
- Customer has multiple projects, their Gradle refers to other repos outside the current repo path (which is ok..)
- We attempt to resolve all the files, and log a warning when we block on we can't

Also, this type of "benign" problem creates too much noise that any signal about genuinely malicious attempts is typically lost.

@viceice @Churro thoughts?

**Comment by viceice** [comment]
we can maybe lower the block message to debug and ignore those files.

**Comment by Churro** [comment]
Changing the logger warning to debug would be my preference too.

It may decrease visibility of genuinely malicious file read attempts but as long as the security control itself remains enforced via `FILE_ACCESS_VIOLATION_ERROR` exception, I don't see an issue. Even with debug, it would still be logged and could explicitly be scanned for.

**Reply by rarkins** [reply]
Thanks, I'll do it now

**Reply by rarkins** [reply]
That brings up a question: do we gracefully handle and move on when such an error is thrown, or do we abort the manager or even whole repo?

**Reply by viceice** [reply]
i think we abort full for now

**Reply by viceice** [reply]
maybe each manager should can a isLocalFile check function before trying to read arbitrary file names extracted from repo

**Reply by Churro** [reply]
What happens in case of exception differs from manager to manager. I randomly traced back some `readLocalFile()` calls:
- composer/extract: Caught in `workers/repository/update/branch/index.ts` -> abort manager
- gradle/parser: Caught in `manager/gradle/extract.ts` -> skip current gradle file
- npm/post-update/pnpm: Caught by `generateLockFile()` (same file) -> skip lockfile update

IMHO, the most predictable behavior would be to simply `return null` with `fs` functions if the check is violated, rather than throwing an exception. I think so because `null` is a legit return value anyway (= failed to read file, write file, etc) and managers tend to already check the return value of `readLocalFile()` and other `fs` functions (at least this holds for the ones I checked).

In that case, it should be enough to simply move `const localFileName = ensureLocalPath(fileName);` inside the `try{}` block of each `fs` function, making it return null in case `ensureLocalPath` throws an exception.

The issue I see with a separate `isLocalFile` check is that it's prone to be overlooked.

**Reply by rarkins** [reply]
I'm happy to return `null`.

BTW ideally we'd also consult `ignorePaths` too before dynamically reading additional files in manager extract!
